### PR TITLE
tinygo: 0.30.0 -> 0.31.1

### PR DESCRIPTION
**Pull request by aykevl**
## Description of changes

Update to version 0.31.0, and simplify the package quite a bit.
Due to changes since 0.30.0, many of the patches can now be dropped.

One functional change is the removal of the wasi-libc dependency: this PR makes sure we build our own version (which is needed, see the comment for details).

~WIP, because 0.31.0 hasn't been released yet. Just wanted to make sure this is available in advance. (There have been a lot of Nix related changes, and I wanted to test them).~

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aykevl** [comment]
I've updated the package to the v0.31.1, which is our last release. It is now ready for review.

File: pkgs/development/compilers/tinygo/0002-buildid.patch
[the remark this answers is not in the document]
Comment (aykevl) [reply pkgs/development/compilers/tinygo/0002-buildid.patch:1]: Yes, thank you! I was able to remove the buildid workaround patch with your patch.
If I understand correctly, all it does is make sure the binary has a valid build ID? That is, it make sure the build ID isn't stripped?

[the remark this answers is not in the document]
Comment (muscaln) [reply pkgs/development/compilers/tinygo/0002-buildid.patch:1]: > If I understand correctly, all it does is make sure the binary has a valid build ID? That is, it make sure the build ID isn't stripped?

Yes, build id is set to empty value by default with #276685.

**Comment by muscaln** [comment]
Result of `nixpkgs-review pr 262346` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tinygo</li>
  </ul>
</details>

**Review by muscaln — APPROVED** [review]
The version in commit message should be `0.31.1`, other than that lgtm.

**Comment by aykevl** [comment]
> The version in commit message should be `0.31.1`

Oops, fixed now.

**Pull request merged**